wordpress-mobile/WordPress-Android
# Unified edit comment component

Part of #15394 , this PR introduce the new Edit Comment Component that, in addition to the comment content, allows to change author parameters (namely: name, web address, email) as it is possible on the web already. Companion Fluxc PR [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2133).

| ![image](https://user-images.githubusercontent.com/47797566/135442660-10bd4736-51a9-438d-9da9-ec0afceb138a.png) | ![image](https://user-images.githubusercontent.com/47797566/135442782-c33357b6-3fa3-401b-a95f-993881777300.png) |
|---|---|

I'm going to add tracking and unit tests in a following PR.

Note 1: the content of the Comment is rendered containing formatting tags as before; we are considering to use GB for comment editing but this is not part of the current scope.

Note 2: end goal is to use this same component both when accessing the Edit functionality from My Site > Comments > Comment Details and Notifications > Comment Details. This PR is introducing it only for the My Site > Comments case; relevant comment data taken from notifications miss or are not always consistent with comments endpoint data, there is an effort ongoing to replace the comment details with a new component that is going to fetch data from comment endpoints also in Notifications scenario for consistency, and we will enable this Edit component for notifications with the new Comment Details component shipping.

## To test
The following should be tested with .COM, JP, pure self-hosted without JP. Double check that the edited comment renders correctly and contains the expected modifications on the web also.

- If the feature flag is not activated, check you get the previous Edit screen both in Notifications > Comment Details and My Site > Comments > Comment Details
- If the feature flag is activated, check that:
   - Check that the previous Edit screen is shown when Editing from Notifications > Comment Details and you can edit the content (smoke test the edited content is saved and reflected on the web as well)
   - Check that the new Edit screen is shown when Editing from My Site > Comments > Comment Details
- For the new component check that:
   - All the fields are rendered as expected
   - You can edit any field and validation occurs based on the field validation function
   - DONE button is active only when at least one field is changed and no validation errors are present
   - DONE button saves the edited comment (double check the web)
   - When no network the DONE operation triggers a snackbar for no network
- Given the Fluxc PR refactor a bit the push function, smock test the comments moderation from My Site > Comments (change comment status to trash, approved, unapproved etc and double check the changes are reflected in the app in the comments list and on the web)

## Regression Notes
1. Potential unintended areas of impact
Feature is still behind feature flag.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Feature is still behind feature flag.

3. What automated tests I added (or what prevented me from doing so)
Feature is still behind feature flag. Going to add unit testing in a following PR.

PR submission checklist:
